compiler/semdata.nim

```diff
@@ -238,8 +249,11 @@ proc makeVarType*(c: PContext, baseType: PType): PType =
     addSonSkipIntLit(result, baseType.assertNotNil)
 
 proc makeTypeDesc*(c: PContext, typ: PType): PType =
-  result = newTypeS(tyTypeDesc, c)
-  result.addSonSkipIntLit(typ.assertNotNil)
+  if typ.kind == tyTypeDesc:
```

Comment:
Supporting these higher-order types is a display of strength, but I'm afraid the rest of the compiler is not there yet. There are a lot of places where various flags and conditions have to be checked more carefully. It's a worthy goal for a future refactoring.